### PR TITLE
stun: add initial app-proto support - v1

### PR DESCRIPTION
**Pull request by jufajardini**
Sharing a first version, so we know if this is the right direction, and also discuss a few things/ TODOs.

Link to ticket: https://redmine.openinfosecfoundation.org/issues/
https://redmine.openinfosecfoundation.org/issues/7068

Describe changes:
- Add initial support for STUN app-proto
- adjust extern C function from STUN to use the new API style
- apply clang format changes to the json schema

### Provide values to any of the below to override the defaults.

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/2016

TODOs:
- remove left-overs that don't make sense for UDP¬
- add docs¬
- add detection (first merge, or later?)¬
- add frames (first merge, or later?)¬

### CODE REVIEW

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on ASAN_TLPR1_suri.
  

Pipeline 22182

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 22241

File: src/output-json-stun.c [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    uint32_t flags;
+    OutputJsonCtx *eve_ctx;
+} LogStunFileCtx;
+
+typedef struct LogStunLogThread_ {
+    LogStunFileCtx *stunlog_ctx;
+    OutputJsonThreadCtx *ctx;
+} LogStunLogThread;
+
+static int JsonStunLogger(ThreadVars *tv, void *thread_data, const Packet *p, Flow *f, void *state,
```
Comment (catenacyber) [inline src/output-json-stun.c:59]: You should not need this output-json-stun file anymore ;-) just do all in rust

File: etc/schema.json [hunk 2]
```diff
@@ -6495,6 +6510,104 @@
             },
             "additionalProperties": false
         },
+        "stun": {
```
Comment (catenacyber) [inline etc/schema.json:6513]: There is no changes in `doc` subdirectory ?

File: rust/src/stun/logger.rs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        js.open_object("response")?;
+        js.set_string("message_type", response.message_type.to_str())?;
+        if let Some(attributes) = &response.attrs {
+            js.open_array("attributes")?;
+            for attribute in attributes {
+                js.start_object()?;
+                if let Some(val) = &attribute.value {
+                    js.set_string_from_bytes(attribute.attr_type.to_str(), val)?;
+                } else {
+        		    js.append_string(attribute.attr_type.to_str())?;
```
Comment (catenacyber) [inline rust/src/stun/logger.rs:39]: Strange to have either a string or a mapping string:string

File: rust/src/stun/parser.rs [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+use nom7::bytes::streaming::take;
+use nom7::combinator::{complete, cond, opt};
+use nom7::multi::{count, many1};
+use nom7::number::streaming::{be_u128, be_u16, be_u8};
+use nom7::IResult;
+
+#[repr(u16)]
+#[derive(Debug, PartialEq, Eq)]
+pub enum StunMethod {
+    Reserved = 0x0000,
```
Comment (catenacyber) [inline rust/src/stun/parser.rs:27]: Maybe the derive EnumString can help

File: rust/src/stun/mod.rs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+pub mod stun;
```
Comment (catenacyber) [inline rust/src/stun/mod.rs:18]: no detection ?

File: rust/src/stun/stun.rs [hunk 6]
```diff
@@ -0,0 +1,511 @@
...
+        get_eventinfo_byid : Some(StunEvent::get_event_info_by_id),
+        localstorage_new: None,
+        localstorage_free: None,
+        get_tx_files: None,
+        get_tx_iterator: Some(applayer::state_get_tx_iterator::<StunState, StunTransaction>),
+        get_tx_data: SCStunGetTxData,
+        get_state_data: SCStunGetStateData,
+        apply_tx_config: None,
+        flags: APP_LAYER_PARSER_OPT_ACCEPT_GAPS,
+        get_frame_id_by_name: None,
```
Comment (catenacyber) [inline rust/src/stun/stun.rs:454]: Can we have frames ?

File: rust/src/stun/stun.rs [hunk 7]
```diff
@@ -0,0 +1,511 @@
...
+
+#[no_mangle]
+pub unsafe extern "C" fn SCStunTxGetAlStateProgress(
+    tx: *mut std::os::raw::c_void,
+    _direction: u8,
+) -> std::os::raw::c_int {
+    let tx = cast_pointer!(tx, StunTransaction);
+
+    // Transaction is done if we have a response.
+    if tx.response.is_some() {
```
Comment (catenacyber) [inline rust/src/stun/stun.rs:411]: We should use `_direction`

File: rust/src/stun/stun.rs [hunk 8]
```diff
@@ -0,0 +1,511 @@
...
+    _rdir: *mut u8
+) -> AppProto {
+    // Need at least 2 bytes.
+    if input_len > 1 && !input.is_null() {
+        let slice = build_slice!(input, input_len as usize);
+        if probe(slice) {
+            return ALPROTO_STUN;
+        }
+    }
+    return ALPROTO_UNKNOWN;
```
Comment (catenacyber) [inline rust/src/stun/stun.rs:305]: need to return error at some point

Is there no fixed pattern for detection ?

File: rust/src/stun/stun.rs [hunk 9]
```diff
@@ -0,0 +1,511 @@
...
+                // For this stun, we'll just try again next time.
+                return AppLayerResult::ok();
+            }
+
+            // It looks like we're in sync with a message header, clear gap
+            // state and keep parsing.
+            self.response_gap = false;
+        }
+        let mut start = input;
+        while !start.is_empty() {
```
Comment (catenacyber) [inline rust/src/stun/stun.rs:238]: Over UDP ? we can have multiple PDUs in one packet ?

File: rust/src/stun/stun.rs [hunk 10]
```diff
@@ -0,0 +1,511 @@
...
+            match parser::parse_message(start) {
+                Ok((rem, response)) => {
+                    start = rem;
+
+                    if let Some(tx) = self.find_request() {
+                        tx.response = Some(response);
+                        SCLogDebug!("Found response for request:");
+                        SCLogDebug!("- Request: {:?}", tx.request);
+                        SCLogDebug!("- Response: {:?}", tx.response);
+                    }
```
Comment (catenacyber) [inline rust/src/stun/stun.rs:248]: create a new tx if request was not seen ?

**Comment by catenacyber** [comment]
Thanks for the work, left a few comments of possible/welcome improvements :-)

File: rust/src/stun/mod.rs (hunk 5 above)
Comment (jufajardini) [reply rust/src/stun/mod.rs:18]: Left that out, as I see some cases in which we first merge the parser, then we move to adding detection capabilities.

File: rust/src/stun/stun.rs (hunk 6 above)
Comment (jufajardini) [reply rust/src/stun/stun.rs:454]: TODO++

File: rust/src/stun/stun.rs (hunk 9 above)
Comment (jufajardini) [reply rust/src/stun/stun.rs:238]: This is (shameful) left-over from using the template to kickstart this development...

**Comment by catenacyber** [comment]
Waiting for the app-layer plugin ;-)

**Pull request closed without merging**